### PR TITLE
Consistently lock and copy ads clients

### PR DESCRIPTION
**Pull request by howardjohn**
fixes https://github.com/istio/istio/issues/28958

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by stevenctl — APPROVED** [review]

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -789,14 +786,13 @@ func (s *DiscoveryServer) startPush(req *model.PushRequest) {
 
 	// Push config changes, iterating over connected envoys. This cover ADS and EDS(0.7), both share
 	// the same connection table
-	s.adsClientsMutex.RLock()
 
 	// Create a temp map to avoid locking the add/remove
-	pending := make([]*Connection, 0, len(s.adsClients))
-	for _, v := range s.adsClients {
+	clients := s.Clients()
+	pending := make([]*Connection, 0, len(clients))
```
Comment (Monkeyanator) [inline pilot/pkg/xds/ads.go:792]: think this should be `s.adsClientsCount()` without the surrounding lock

Comment (Monkeyanator) [reply pilot/pkg/xds/ads.go:792]: ah ignore that, misread the line 👍

Comment (howardjohn) [reply pilot/pkg/xds/ads.go:792]: No I think you are right? We are copying it here, but Clients already copies

File: pilot/pkg/xds/debug.go [hunk 2]
```diff
@@ -448,18 +444,12 @@ func (s *DiscoveryServer) adsz(w http.ResponseWriter, req *http.Request) {
 	w.Header().Add("Content-Type", "application/json")
 	if req.Form.Get("push") != "" {
 		AdsPushAll(s)
-		s.adsClientsMutex.RLock()
-		_, _ = fmt.Fprintf(w, "Pushed to %d servers", len(s.adsClients))
-		s.adsClientsMutex.RUnlock()
+		_, _ = fmt.Fprintf(w, "Pushed to %d servers", len(s.Clients()))
```
Comment (Monkeyanator) [inline pilot/pkg/xds/debug.go:447]: doing `s.adsClientCount()` here would avoid allocation in `Clients()` I think? (not sure it matters in practice)

Comment (howardjohn) [reply pilot/pkg/xds/debug.go]: Probably doesn't matter since its a debug endpoint, but might as well make it more optimized. thanks

**Review by hzxuzhonghu — APPROVED** [review]

**Review by ramaraochavali — APPROVED** [review]

**Pull request merged**